enter.c

```diff
@@ -153,6 +153,35 @@ int enter(struct entry_settings *opts)
 		reset_capabilities();
 	}
 
+	const char *root = (opts->root != NULL) ? opts->root : "/";
+
+	pid_t rootpid = getpid();
+
+	// If this set (!= -1), then it is the open fd of bst directory needed to be cleaned
+	int cleanfd = -1;
+
+	/*
+	 * If there is a cgroup specified we need to create bst.<pid> subcgroup to join
+	 * Otherwise if no cgroup is specified but limits are applied bst will error out
+	 */
+	if (opts->cgroup_path != NULL) {
+		cleanfd = open(opts->cgroup_path, O_PATH | O_DIRECTORY | O_CLOEXEC);
```

Comment:
This is a bit delicate. All of the code prior to `setuid(getuid())` on L207 is privileged, and may be running as setuid root. This means that the open will always succeed.

The only way this can be secure is if you either open the file after dropping privileges and then send the file descriptor to the helper, or leave the current code as-is and perform an explicit permission check. The easier (and more correct) way to do this is the former, since the latter requires you to emulate the kernel permission model, and we're going to get it wrong (checking user & mode is insufficient, and ignores acls, selinux and apparmor, seccomp, ...).

I have some pending changes to allow this as part of my seccomp supervisor. I'll cherry-pick some of the commits that does this, and you'll just be able to call send_fd.

Reply:
Added setsid() to cgroup_helper.